src/ws/cockpithandlers.c

```diff
@@ -131,6 +121,99 @@ cockpit_handler_socket (CockpitWebServer *server,
   return TRUE;
 }
 
+gboolean
+cockpit_handler_external (CockpitWebServer *server,
+                          const gchar *path,
+                          GIOStream *io_stream,
+                          GHashTable *headers,
+                          GByteArray *input,
+                          guint in_length,
+                          CockpitHandlerData *ws)
+{
+  CockpitWebResponse *response = NULL;
+  CockpitWebService *service = NULL;
+  const gchar *segment = NULL;
+  JsonObject *open = NULL;
+  const gchar *query = NULL;
+  CockpitCreds *creds;
+  const gchar *expected;
+  const gchar *upgrade;
+  guchar *decoded;
+  GBytes *bytes;
+  gsize length;
+
+  /* The path must start with /cockpit+xxx/channel/csrftoken? or similar */
+  if (path && path[0])
+    segment = strchr (path + 1, '/');
+  if (!segment)
+    return FALSE;
+  if (!g_str_has_prefix (segment, "/channel/"))
+    return FALSE;
+  segment += 9;
+
+  /* Make sure we are authenticated, otherwise 404 */
+  service = cockpit_auth_check_cookie (ws->auth, path, headers);
+  if (!service)
+    return FALSE;
+
+  creds = cockpit_web_service_get_creds (service);
+  g_return_val_if_fail (creds != NULL, FALSE);
+
+  expected = cockpit_creds_get_csrf_token (creds);
+  g_return_val_if_fail (expected != NULL, FALSE);
+
+  /* The end of the token */
+  query = strchr (segment, '?');
+  if (!query)
+    query = segment + strlen (segment);
+
+  /* No such path is valid */
+  if (strncmp (expected, segment, query - segment) != 0)
```

Comment:
This matches when `segment` is a prefix of `expected`, including when `segment` is the empty string.  We also need to check `strlen(expected) == query - segment`.